### DISCUSSION TITLE
how can one draw a polygon on the map at given coordinates

### DISCUSSION THREAD

**Discussion by stleusc** [mapsforge/vtm#845] [Q&A] [answered]
I see examples like the HexagonRenderer however all example I find simply draw on the center of the map.
I want to draw at give lat/lon points. Also don't want to do tiling like the default VectorTileRenderer.

A simple question would be, how to modify HexagonRenderer example file to render the hexagons centered at a give lat/lon (not middle of map) and size them a give size in meters/feet on the map surface to make sure it scales.

For simplicity, what would need to be done to draw the HexagonRenderer and N8 00.000 W5 00.000 with a size of 100m per hexagon. Drawing a single Hexagon centered at that point is good enough as an example.

It can't be too hard but not finding any good starting point.

Thanks everyone!

**Comment by devemux86** [comment]
You can see an example with overlays, like [VectorLayerActivity](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/VectorLayerActivity.java).

It can be adapted to overlay various geometries with `vtm-jts`.

**Reply by stleusc** [reply]
Thanks!
I had seen that also but had, actually still have some trouble with it!

```java
            Style style = Style
                    .builder()
                    .strokeWidth(Util.dpToPx(2, getContext()))
                    .strokeColor(Color.RED)
                    .fillAlpha(0.5f)
                    .fillColor(Color.RED)
                    .build();

            layer.add(new PolygonDrawable((List<GeoPoint>)getPoints(), style));
```

NOTHING I am trying gives me a FILLED polygon!
I do see the polygon on the map, red outline, 2dp wide. TRANSPARENT fill!
I picked alpha in the middle in case somethings was inverted. But also 1.0f makes no difference.

Latest repo code used...
Any idea why it would only do outline?

**Comment by devemux86** [comment, marked as answer]
The following code seems to work:

```java
Style style = Style.builder()
        .fillAlpha(0.3f)
        .fillColor(Color.BLUE)
        .strokeColor(Color.BLUE)
        .strokeWidth(2)
        .build();
vectorLayer.add(new PolygonDrawable(style,
        new GeoPoint(-20d, -20d),
        new GeoPoint(20d, -20d),
        new GeoPoint(20d, 20d),
        new GeoPoint(-20d, 20d)));
```

![VTM](https://user-images.githubusercontent.com/3484020/118937701-766b4680-b956-11eb-8af8-d27269fb0e54.png)

**Reply by stleusc** [reply]
Something is funky on my end. 
I know I am using a rather old codebase (I heavily had extended some of the features). I plan to migrate one day (soon, lol).

So I don't expect too much help on this, I understand if you can't or don't want to support this much. So your help is much appreciated,

Having buildings drawn or other features I of course know that filling is working in general.
For testing I removed all map layers and only add the single square above.
And it is outlined only!

Maybe you can help me by advising how I could best debug the drawing?
I'm sure it's likely not a bug in the current code base but for historical reasons I can not just jump now.

VectorLayer and Geometry are the latest!
Makes me wonder.... Renderscripts maybe? I'll check that next.

I simply need to get it to work with my version for now.

**Reply by stleusc** [reply]
updated RenderBucket(s) and PolygonBucket and all shaders.
nothing so far. wondering what else is involved in rendering the FILL?

**Reply by demiantres** [reply]
The VTM code does not correctly handle "enableVertexAttribArray()" is some situations meaning that it can stop to draw correctly when you do some drawing outside of VTM (see https://github.com/mapsforge/vtm/discussions/829). You can try to manually call enableVertexAttribArray before each rendering loop to check if it fixes the problem.

**Reply by stleusc** [reply]
@demiantres thanks for that comment. It seems like it is not directly related to my issue though!

**Reply by demiantres** [reply]
I mentioned it because the possible bug in my linked thread also caused polygons in vector tiles to sometimes not load correctly. You can quickly test if your problem is related to this by just enabling all vertex arrays at the beginning of the rendering loop.

**Comment by devemux86** [comment]
The general advice in such cases is usually:

If the upstream repository works correctly, while you use old library and the custom code does not work,
then you need to update the library and check / compare the code with the upstream to find if there are
any new upstream improvements or some error in the custom modifications.

**Reply by stleusc** [reply]
I thought I explained my reasons. Telling me what I have to do while I explained that I currently can not isn't really helpful!
One can NOT always just switch. I did not expect anyone to solve my problem. More for pointers to debug.

Here what I found...

Below code is not really called on polygons in my case.
```java
        t.buckets.prepare();
```

By ensuring prepare is called on the bucket I got the fill to work.

Like I said, I did not imply a bug in the current code base and it seems like it got fixed somewhere, somehow, sometime.

**Comment by stleusc** [comment]
AbstractVectorLayer.java

```java
		/** running on worker thread */
		@Override
		public boolean doWork(Task t) {
			Viewport v = mMap.viewport();
			BoundingBox bbox;
			synchronized (v) {
				bbox = v.getBBox();
				v.getMapPosition(t.position);
			}

			processFeatures(t, bbox);

			t.buckets.prepare(); // missing for me, now present in current version....

			mMap.render();
			return true;
		}

	}
```

Above my fix that is needed for MY version. Repo code now has that line added. My copy was missing the prepare() statement.

Thanks everyone, Can be closed now.

**Comment by devemux86** [comment]
> Repo code now has that line added. My copy was missing the prepare() statement.

The fork has always had this line from the beginning.
(it was added in 2015)

**Reply by stleusc** [reply]
I told you, old code base ;-) VERY OLD CODEBASE. Which is why it is hard to switch. But I will eventually :-)

@devemux86 thanks for your support on this odd topic!
